### ISSUE TITLE
[MPS] softmax returns NaN attention probabilities for large tensors, in float16 and float32.

### ISSUE THREAD

**Issue by Birch-san** [pytorch/pytorch#96602]
### 🐛 Describe the bug

I've been making large images in stable-diffusion in float16 on my 4090 for a while.  
I tried on my M1 Max, generating an image at a size which _is_ possible in float16 on CUDA.

Using a recent PyTorch nightly (`2.1.0.dev20230310`), macOS 13.3 public beta.  
fails in float32 also.

```python
from torch import full, softmax, FloatTensor
import torch

# fill value is based on mean() from some real attention scores I had
attention_scores: FloatTensor = full([10, 12416, 12416], 0.0598, device='mps', dtype=torch.float16)

attention_probs: FloatTensor = attention_scores.softmax(dim=-1)

assert not attention_probs.isnan().any().item(), "found NaN in attention_probs"
# interestingly, attention_scores *also* becomes NaN. does softmax() mutate the input?
```

to pinpoint _where_ in the softmax it fails: here's a decomposed softmax.

```python
from torch import full, FloatTensor
import torch

attention_scores: FloatTensor = full([10, 12416, 12416], 0.0598, device='mps', dtype=torch.float16)

def softmax(x: FloatTensor, dim=1) -> FloatTensor:
  maxes = torch.max(x, dim, keepdim=True).values
  assert not maxes.isnan().any().item(), "found NaN in maxes"
  diffs = x-maxes
  assert not diffs.isnan().any().item(), "found NaN in diffs" # this assertion gets tripped
  x_exp = torch.exp(diffs)
  assert not x_exp.isnan().any().item(), "found NaN in x_exp"
  x_exp_sum = torch.sum(x_exp, dim, keepdim=True)
  assert not x_exp_sum.isnan().any().item(), "found NaN in x_exp_sum"
  quotient = x_exp/x_exp_sum
  assert not quotient.isnan().any().item(), "found NaN in quotient"
  return quotient

attention_probs: FloatTensor = softmax(attention_scores, dim=-1)
assert not attention_probs.isnan().any().item(), "found NaN in attention_probs"
```

interestingly, the NaNs are introduced when computing `diffs = x-maxes`. even though this really should just be computing lots of `0.0598-0.0598=0`s.

I have exported a realistic `attention_scores` tensor, but it's 3GB. not sure how to share a file that large. tell me if it'd be useful to have.

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230310
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.3 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.22.1
Libc version: N/A

Python version: 3.11.0 | packaged by conda-forge | (main, Jan 14 2023, 12:26:40) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.3-arm64-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] clip-anytorch==2.5.2
[pip3] numpy==1.24.2
[pip3] torch==2.1.0.dev20230310
[pip3] torchdiffeq==0.2.3
[pip3] torchsde==0.2.5
[pip3] torchvision==0.15.0.dev20230310
[conda] clip-anytorch             2.5.2                    pypi_0    pypi
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch                   2.1.0.dev20230310        py3.11_0    pytorch-nightly
[conda] sympy                     1.11.1                  py311_0    pytorch-nightly
[conda] torchdiffeq               0.2.3                    pypi_0    pypi
[conda] torchsde                  0.2.5                    pypi_0    pypi
[conda] torchvision               0.15.0.dev20230310          pypi_0    pypi
```

**Comment by Parskatt** [comment]
@malfet This still remains.

**Comment by albanD** [comment]
@drisspg is the the math vs fused precision issue we were discussing?

**Comment by drisspg** [comment]
I think this is likely pretty similar to https://github.com/pytorch/pytorch/issues/121558.. FMA is the evil lol

**Comment by rsamerica** [comment]
**Bug still reproduces in 2026 on much newer torch + macOS — possibly hardware-specific to M5 Max GPU**

Following up on the most recent comment (non-repro on torch 2.5.1 + macOS 15.1): I can confirm the bug is **still active** on substantially newer software, with hardware as the apparent variable.

**Reproduction:**
- **Hardware:** Apple M5 Max, 128 GB unified memory
- **OS:** macOS 26.4.1
- **PyTorch:** Tested both `2.10.0` (stable) and `2.13.0.dev20260424` (nightly) — both reproduce
- **Workload:** Stable Diffusion XL inference via ComfyUI 0.19.5

**Hardware-regression observation (potentially the key variable):** Same install, same Python venv, same workflows, same checkpoints — bug was **not** present on M4 Pro before Migration Assistant transfer to M5 Max. Pure hardware change → bug returns. This is consistent with both the maintainer's hypothesis that an OS- or library-side fix exists *and* with the bug being specific to GPU dispatch paths that M5's new GPU architecture exercises differently. If the non-repro on macOS 15.1 was on M1/M2/M3, M5 may simply have brought it back.

**Symptom:** SDXL UNet sampling on MPS intermittently produces dead/NaN latents that decode as fully black images. Failure rate ~15–50% depending on resolution (1024² stabler than 768×1344). First generation after server start usually succeeds; subsequent generations fail at increasing rate, suggesting MPS allocator/cache state accumulation. Full app restart temporarily resets it.

**Architecture A/B (still pointing at softmax/attention):** A DiT-based diffusion model (Z-Image Turbo) running in the same ComfyUI process on the same hardware at identical resolutions produces zero failures across many runs. Bug remains specific to SDXL's UNet attention path — consistent with the original softmax-NaN diagnosis.

**Workarounds tried, none effective (consistent with the original report's finding that fp32 is also affected):**
- `--fp32-vae`
- `--force-upcast-attention` (forces FP32 attention math)
- `--use-split-cross-attention` (alternate attention path)
- `--cpu-vae` — proves bug is upstream of VAE; sampling itself produces dead latents
- bf16 VAE (auto-selected by torch nightly)
- Multiple SDXL checkpoints, including freshly-downloaded weights

**Suggestion:** Worth keeping this issue open. The "fixed on torch 2.5.1 + macOS 15.1" non-repro may have been GPU-specific (M-series before M5). If a maintainer can repro on M5 Max — even briefly — we'd have a concrete bisection target.

Happy to provide a minimal standalone repro (raw safetensors UNet load + N forward passes through PyTorch only, no ComfyUI/diffusers dependency) on M5 if useful. Let me know.

**Comment by jhavukainen** [comment]
Hi @rsamerica please share the minimal standalone repro if you have it. I can take a look since I'm on the topic of M5 specific issues right now.

**Comment by rsamerica** [comment]
**Update — narrowed the trigger; may be a different mechanism than this issue tracks**

Following up on my earlier M5 Max comment. I attempted to reduce to a minimal repro, and the result reframes the diagnosis.

**What does NOT reproduce on M5 Max (torch 2.10.0, macOS 26.4.1):**
- Single SDXL UNet forward with random inputs (32 trials, 0 failures)
- Full Euler sampling loop with CFG, 30 steps, random embeddings (8 trials, 0 failures)
- Same at 768×1344 (the originally failing resolution) — 8 trials, 0 failures
- Full `StableDiffusionXLPipeline` — UNet + VAE + both CLIP encoders resident on MPS, real prompts, decoded and saved as PNGs — 8 trials, 0 failures

512+ UNet forward passes on the bug's worst-case resolution with full pipeline, all clean. Standalone diffusers on M5 Max does **not** reproduce.

**What DOES reproduce: ComfyUI specifically.** Same hardware, same torch, same checkpoint, same resolution. The differentiator is ComfyUI's `smart_memory` feature, which performs model swap operations between sampling and VAE decode. The feature is designed to relieve VRAM pressure on CUDA hardware (separate VRAM/RAM); on Apple Silicon's unified memory there is no separate pool to relieve, so the swaps are gratuitous churn — but they still exercise MPS's caching allocator paths.

**Clean A/B with `--disable-smart-memory`:**

| Smart memory | Trials | Failures | Rate |
|---|---|---|---|
| OFF | 48 | 0 | 0.0% |
| ON | 16 | 4 | 25.0% |

Same hardware, same SDXL checkpoint, same resolution (768×1344), same step/CFG. One toggle flipped. Fisher's exact p < 10⁻⁵.

**Failure-mode signature** (from ComfyUI stderr on a black-image generation):

```
30/30 [00:21<00:00, 1.39it/s]
.../nodes.py:1662: RuntimeWarning: invalid value encountered in cast
  img = Image.fromarray(np.clip(i, 0, 255).astype(np.uint8))
```

`np.clip(NaN, 0, 255).astype(np.uint8) → 0 → black`. NaN is present in the image tensor before the cast, and (since CPU VAE was previously confirmed not to help) it's already present in the latent before VAE decode. The sampler is producing NaN latents under specific MPS allocator states.

**Hypothesis:** the bug is in MPS allocator behavior when buffers are freed and reallocated across model swap cycles. Off-square latent shapes (768×1344, 832×1216, 896×1216) amplify the failure rate vs. square 1024×1024 — possibly because off-square shapes exercise kernel paths less robust to dirty buffer reuse.

**This may not be the same mechanism as the softmax NaN this issue tracks.** Tonight's findings point at allocator state, not deterministic math. Cross-posting the workaround at Comfy-Org/ComfyUI#10681 (an existing report with the same warning signature on Mac Studio M3 Ultra 256GB that's been sitting open since November). If a more surgical PyTorch-side repro (a `.to("cpu")` / `.to("mps")` cycle script that triggers dirty buffer reuse) would help, I can write one — let me know.

Tested only on M5 Max + 128 GB unified RAM. Whether the same allocator behavior affects smaller-RAM Apple Silicon configurations is not determined.

**Comment by qwqsilver00-lang** [comment]
Adding another **M5 Max** reproduction (2026-08), with a black-video artifact attached.

**Environment**
- Apple M5 Max, 128 GB unified memory
- macOS 26.6 (Tahoe)
- PyTorch `2.13.0.dev20260518` (also reproduces on stable 2.10.0)
- Backend: MPS

**Symptom**
Diffusion-model inference in **bfloat16** intermittently produces NaN attention probabilities → NaN latents → completely black output. The attached video (`LTX23_00110-audio.mp4`) is a real black-frame output from an LTX-Video run on M5 Max in bf16. The decode step logs:

```
videohelpersuite/nodes.py:130: RuntimeWarning: invalid value encountered in cast
  return tensor_to_int(tensor, 8).astype(np.uint8)
```

i.e. the decoded tensor contains NaN/inf → `.astype(np.uint8)` → 0 → black.

**What works / what doesn't**
- bf16 model (default on Apple Silicon) → black, intermittent
- `--force-fp32` (whole model fp32) → correct, but ~2× memory and ~2× slower per step
- `--fp32-vae` only → still black (VAE is not the source; NaN is already in the latent)
- `--use-pytorch-cross-attention` (SDPA instead of sub-quadratic baddbmm) → still black on this model

So the NaN is produced during bf16 diffusion sampling, consistent with @drisspg's FMA-precision hypothesis (#121558) and the `x - maxes` failure mode in the original report.

**Hardware regression note**
Same as @rsamerica's observation above: this was correct on M4-class hardware; M5's GPU re-exposes the NaN. The 2025-01 "fixed on macOS 15.1" did not hold for M5.

@jhavukainen — happy to provide a minimal standalone repro or the full ComfyUI log if useful. The community thread with many affected M-series users is Comfy-Org/ComfyUI#15315.

https://github.com/user-attachments/assets/f399de6f-8b82-4962-a1ca-6fc6d8922e9d
